Services/UICore/classes/class.ilGlobalTemplate.php

```diff
@@ -0,0 +1,1998 @@
+<?php
+/* Copyright (c) 1998-2009 ILIAS open source, Extended GPL, see docs/LICENSE */
+
+include_once("./Services/UICore/lib/html-it/IT.php");
+include_once("./Services/UICore/lib/html-it/ITX.php");
+
+/**
+* special template class to simplify handling of ITX/PEAR
+* @author	Stefan Kesseler <skesseler@databay.de>
+* @author	Sascha Hofmann <shofmann@databay.de>
+* @version	$Id$
+*/
+class ilGlobalTemplate
+{
+	
+
+	
+	protected $tree_flat_link = "";
+	protected $page_form_action = "";
+	protected $permanent_link = false;
+	
+	protected $lightbox = array();
+	protected $standard_template_loaded = false;
+
+	protected $translation_linked = false; // fix #9992: remember if a translation link is added
+
+	/**
+	 * @var	\ilTemplate
+	 */
+	protected $template;
+
+	/**
+	* constructor
+	* @param	string	$file 		templatefile (mit oder ohne pfad)
+	* @param	boolean	$flag1 		remove unknown variables
+	* @param	boolean	$flag2 		remove empty blocks
+	* @param	boolean	$in_module	should be set to true, if template file is in module subdirectory
+	* @param	array	$vars 		variables to replace
+	* @access	public
+	*/
+	public function __construct($file,$flag1,$flag2,$in_module = false, $vars = "DEFAULT",
```

Comment:
In my understanding the global template is the combination of tpl.main.html with tpl.adm_content.html included.

- Everytime `tpl.main.html` or `tpl.adm_content.html` is used in consumer code this is evil and should be replaced by calling loadStandardTemplate (see my comment below)
- Everytime `loadStandardTemplate` is called in consumer code, this is a use of the global template, so a `new ilGlobalTemplate` is necessary there (and I would expect only in these cases).
- I would expect that we won't need the `$file, $flag1, ...` parameters in ilGlobalTemplate anymore, since they only handle the combination of tpl.main.html / tpl.adm_content.html.
- We could get rid of the `loadStandardTemplate` calls in the last step, since `ilGlobalTemplate` could do this automatically.

Does this make sense?

Reply:
Yes this indeed makes sense, but one problem currently imo stands in the way of proceeding as you suggest (or I am missing something): I found locations (like SCORM) where `ilGlobalTemplate` is initialized via `new` and then later on either `printToStdout` or `getSpecial` is used on that instance. These methods in turn use a globals, require special variables in a template, use other internals of ilGlobalTemplate, etc. So I think we are seeing two different responsibilities in the class that I could not disentagle until now: One is summarized by your understanding (tpl.main.html + tpl.adm_content.html), the other is a lot messier and currently is even hard to describe for me in a catchy sentence. `printToStdOut` and `getSpecial` together with the fill*-methods seem to encode a set of common variables and blocks in the templates as well as a pattern to render them and use globals to render them. I am very afraid of touching that.

I will add a dicto rule that forbids the use of `tpl.main.html` and `tpl.adm_content.html` anywhere besides `ilGlobalTemplate`, but I currently cannot see how to get rid of the other usages of `ilGlobalTemplate` (besides the "copy"-move).

Reply:
What we could do then is split `ilGlobalTemplate` again, where one automatically loads "tpl.main.html" (and has no parameters to the constructor) and one encodes the messy rendering order. The latter then would be a baseclass of the former. Should I do that?

Reply:
This sounds good IMO.

Reply:
ilMainTemplate? Yes I think this should be feasible.